temporal-test-server/src/main/java/io/temporal/internal/testservice/TestWorkflowService.java

```diff
@@ -1054,6 +1063,171 @@ public void pollWorkflowExecutionUpdate(
+                .setMessage("INVALID_ARGUMENT: RequestEagerExecution is not supported.")
+                .build(),
+            null);
+      }
+
+      UpdateWorkflowExecutionRequest updateRequest;
+      ExecuteMultiOperationRequest.Operation secondOperation = request.getOperations(1);
+      if (secondOperation.getOperationCase() != UPDATE_WORKFLOW) {
+        throw Status.INVALID_ARGUMENT
+            .withDescription("Operations have to be exactly [Start, Update].")
+            .asRuntimeException();
+      }
+      updateRequest = secondOperation.getUpdateWorkflow();
+
+      if (!updateRequest.getWorkflowExecution().getRunId().isEmpty()) {
+        throw multiOperationExecutionFailure(
+            null, // start aborted
+            MultiOperationExecutionFailure.OperationStatus.newBuilder()
+                .setCode(Status.INVALID_ARGUMENT.getCode().value())
+                .setMessage("INVALID_ARGUMENT: RunId is not allowed.")
+                .build());
+      }
+
+      if (!updateRequest.getFirstExecutionRunId().isEmpty()) {
+        throw multiOperationExecutionFailure(
+            null, // start aborted
+            MultiOperationExecutionFailure.OperationStatus.newBuilder()
+                .setCode(Status.INVALID_ARGUMENT.getCode().value())
+                .setMessage("INVALID_ARGUMENT: FirstExecutionRunId is not allowed.")
+                .build());
+      }
+
+      if (!startRequest
+          .getWorkflowId()
+          .equals(updateRequest.getWorkflowExecution().getWorkflowId())) {
+        throw multiOperationExecutionFailure(
+            null, // start aborted
+            MultiOperationExecutionFailure.OperationStatus.newBuilder()
+                .setCode(Status.INVALID_ARGUMENT.getCode().value())
+                .setMessage(
+                    "INVALID_ARGUMENT: WorkflowId is not consistent with previous operation(s)")
+                .build());
+      }
+
+      if (startRequest.hasWorkflowStartDelay()) {
+        throw multiOperationExecutionFailure(
+            MultiOperationExecutionFailure.OperationStatus.newBuilder()
+                .setCode(Status.INVALID_ARGUMENT.getCode().value())
+                .setMessage("INVALID_ARGUMENT: WorkflowStartDelay is not supported.")
+                .build(),
+            null);
+      }
+
+      // execute start
+      StartWorkflowExecutionResponse startResult;
+      try {
+        startResult =
+            startWorkflowExecutionImpl(
+                startRequest, Duration.ZERO, Optional.empty(), OptionalLong.empty(), null);
+      } catch (StatusRuntimeException e) {
+        throw multiOperationExecutionFailure(
+            MultiOperationExecutionFailure.OperationStatus.newBuilder()
+                .setCode(e.getStatus().getCode().value())
+                .setMessage(e.getMessage())
+                .build(),
+            null); // update aborted
+      }
+
+      // execute update
+      UpdateWorkflowExecutionResponse updateResult;
```

Comment:
How do you know this update is always sent as part of the first workflow task?

Reply:
The test is good, but looking at the code it is not clear to me the test server implementation actually **guarantees** the update will be in the same workflow task? My specific concern is the  `startWorkflowExecutionImpl` and the `updateWorkflowExecutionImpl` are separate calls?

Reply:
I checked out your branch and added a short 1s sleep between the call `startWorkflowExecutionImpl ` and `updateWorkflowExecutionImpl ` and confirmed the update will not be delivered in the first workflow task

Reply:
Okay, I've poked around the server a little more, in particular `TestWorkflowMutableStateImpl`, and can see how this happens now. I don't see an easy fix right now. The two MS updates need to happen together to guarantee they are also delivered together to the worker. But I don't see any kind of "transaction" API. I'll need to dig deeper.

Reply:
Addressed here: https://github.com/temporalio/sdk-java/pull/2199/commits/2a39d5edb5b73ed5a9cb976fe5038105d7ea2f1c